pkg/kubeapiserver/authorizer/config.go

```diff
@@ -141,3 +158,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
 
 	return union.New(authorizers...), union.NewRuleResolvers(ruleResolvers...), nil
 }
+
+type ruleResolvedAUthorizer interface {
+	Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error)
+	RulesFor(user user.Info, namespace string) ([]authorizer.ResourceRuleInfo, []authorizer.NonResourceRuleInfo, bool, error)
+}
+
+type withNotSyncedReason struct {
+	ruleResolvedAUthorizer
+	informers []cache.SharedInformer
+}
+
+func (az withNotSyncedReason) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	d, reason, err := az.ruleResolvedAUthorizer.Authorize(ctx, a)
+	if err != nil {
+		return d, reason, err
+	}
+
+	if d == authorizer.DecisionNoOpinion && len(reason) == 0 && !az.hasSynced() {
+		reason = "informers not synced"
+		klog.Warningf("authorize before synced informers by %s for %s at %s", a.GetUser().GetName(), a.GetVerb(), a.GetPath())
+	}
+
+	return d, reason, err
+}
+
+func (az withNotSyncedReason) hasSynced() bool {
+	for _, i := range az.informers {
+		if !i.HasSynced() {
+			return false
+		}
+	}
+
+	return true
+}
```

Comment:
This seems like an expensive check to add for every no-opinion response the wrapped authorizers will ever make. Wherever we keep HasSynced checks in these authorizer critical paths, it seems like they should be made ratcheting, so once they pass once it remembers and never has to iterate / recursively lock multiple informers

Stepping back, wrapping authorizers this way can lead to some misleading reasons being output (like the node authorizer reporting a rejecting reason for requests from non-node users)

Would it make more sense to work to minimize how often this is user-visible by making the authorizers that depend on informer syncing (rbac and node) do the hasSynced check internally and hold a relevant request for 
a short time (~1 second) for the first n seconds of their lifetime if they have not yet synced?

Also, there have been other attempts to include global informer synced state in API server readiness to avoid routing requests to servers not yet prepared to handle them. cc @wojtek-t

Reply:
actually I think we have that already from https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L257. It might be that we are hunting early requests to an unread kube-apiserver here.